packages/headless/src/features/case-assist/case-assist-analytics-selectors.test.ts

```diff
@@ -0,0 +1,200 @@
+import {
+  KnownCaseFields,
+  selectCase,
+  selectCaseClassification,
+  selectCaseFieldValue,
+  selectCaseInputValue,
+  selectCustomCaseFieldValues,
+  selectCustomCaseInputValues,
+} from './case-assist-analytics-selectors';
+
+describe('case assist analytics selectors', () => {
+  const stateWithCaseInputs: Partial<CaseAssistAppState> = {
+    caseInput: {
+      id: {value: 'some-case-id'},
+      subject: {value: 'some case subject'},
+      description: {value: 'some case description'},
+      productId: {value: 'some product ID'},
+      customFieldA: {value: 'custom value'},
+    },
+  };
+
+  const stateWithCaseFields: Partial<CaseAssistAppState> = {
+    caseField: {
+      status: {
+        loading: false,
+        error: null,
+        lastResponseId: 'last-response-id',
+      },
+      fields: {
+        category: {
+          value: 'software',
+          suggestions: [
+            {
+              id: 'software-suggestion-id',
+              value: 'software',
+              confidence: 0.987,
+            },
+            {
+              id: 'service-suggestion-id',
+              value: 'service',
+              confidence: 0.765,
+            },
+          ],
+        },
+        customFieldB: {
+          value: 'custom value',
+          suggestions: [],
+        },
+      },
+    },
+  };
+
+  describe('#selectCase', () => {
+    it('should retrieve all the case information from the state', () => {
+      const caseInfo = selectCase({
+        ...stateWithCaseInputs,
+        ...stateWithCaseFields,
+      });
+
+      expect(caseInfo).toMatchObject({
+        id: 'some-case-id',
+        subject: 'some case subject',
+        description: 'some case description',
+        productId: 'some product ID',
+        category: 'software',
+        custom: {
+          customFieldA: 'custom value',
+          customFieldB: 'custom value',
+        },
+      });
+    });
+  });
+
+  describe('#selectCaseInputValue', () => {
+    it('should return the input value', () => {
+      const value = selectCaseInputValue(stateWithCaseInputs, 'subject');
+
+      expect(value).toBe('some case subject');
+    });
+
+    it('should return #undefined when field is not found', () => {
+      const value = selectCaseInputValue(
+        stateWithCaseInputs,
+        'thisFieldDoesNotExist'
+      );
+
+      expect(value).toBeUndefined();
+    });
+  });
+
+  describe('#selectCustomCaseInputValue', () => {
+    it('should return all inputs except the known fields', () => {
+      const customFields = selectCustomCaseInputValues(stateWithCaseInputs);
+
+      expect(customFields).toHaveProperty('customFieldA', 'custom value');
+      Object.values(KnownCaseFields).forEach((knownCaseField) =>
+        expect(customFields).not.toHaveProperty(knownCaseField)
+      );
+    });
+
+    it('should return an empty object if no custom fields', () => {
+      const state = JSON.parse(JSON.stringify(stateWithCaseInputs));
```

Comment:
Why the stringify then parse ? Isn't that just giving you back the same object ? 🤔 

Is it to obtain a copy ? If so, you could do 
```
const state = {...stateWithCaseInputs}
```

Reply:
Yes, the idea is to clone the object. Unfortunately the spread operator only work for the first level properties, the object references are still used for deeper levels. Altering `stateWithCaseInputs` could cause side-effects with other tests.

I can add a local "clone" function just to make the intent clearer.

Reply:
Another approach would be to have a function that return a new "stateWithCaseFields" or "stateWithCaseInputs" instead.

As you prefer !

Reply:
You're right, using a method to build the state object is better.